consensus/parlia/snapshot.go

```diff
@@ -211,17 +218,45 @@ func (s *Snapshot) updateAttestation(header *types.Header, chainConfig *params.C
 	}
 }
 
-func (s *Snapshot) SignRecently(validator common.Address) bool {
+func (s *Snapshot) versionHistoryCheckLen() uint64 {
+	return uint64(len(s.Validators)) * uint64(s.TurnLength)
+}
+
+func (s *Snapshot) minerHistoryCheckLen() uint64 {
+	return (uint64(len(s.Validators))/2+1)*uint64(s.TurnLength) - 1
+}
+
+func (s *Snapshot) countRecents() map[common.Address]uint8 {
+	leftHistoryBound := uint64(0) // the bound is excluded
+	checkHistoryLength := s.minerHistoryCheckLen()
+	if s.Number > checkHistoryLength {
+		leftHistoryBound = s.Number - checkHistoryLength
+	}
+	counts := make(map[common.Address]uint8, len(s.Validators))
 	for seen, recent := range s.Recents {
-		if recent == validator {
-			if limit := uint64(len(s.Validators)/2 + 1); s.Number+1 < limit || seen > s.Number+1-limit {
-				return true
-			}
+		if seen <= leftHistoryBound || recent == (common.Address{}) {
```

Comment:
`seen < leftHistoryBound`? equal to leftHistoryBound would be seen as recents;
take it for example:
```
s.TurnLength = 1
len(s.Validators) = 21
minerHistoryCheckLen() = 10
s.Number = 110
leftHistoryBound = 100
```
in this case block number 100 would be counted as one of the recent sign

Reply:
`recent == (common.Address{})`  maybe we can add a warning log, as it is abnormal.

Reply:
> `recent == (common.Address{})` maybe we can add a warning log for it, as it is abnormal.

when seen == `epochKey`
I will update the comment in codes

Reply:
> `seen < leftHistoryBound`? equal to leftHistoryBound would be seen as recents; take it for example:
> 
> ```
> s.TurnLength = 1
> len(s.Validators) = 21
> minerHistoryCheckLen() = 10
> s.Number = 110
> leftHistoryBound = 100
> ```
> 
> in this case block number 100 would be counted as one of the recent sign

no, 100 is excluded
(101,110] is the range, and the length is 10